### ISSUE TITLE
After heavy reads on zvols and datasets, the system hangs with 1.4.5

### ISSUE THREAD

**Issue by vmrob** [openzfsonosx/zfs#419]
I noticed that my zfs info scripts were starting to time out and so I ran a spindump before the whole system locked up. When I say that the system locks up, I should be really specific: I'm still able to do everything I would normally do, but as soon as any process attempts to write to disk (save for a few finder operations) and, in some cases, read from disk, the beachball of death starts and never finishes. I'm generally unable to start new processes on the command line as terminal has definitely locked up and I typically can't ssh into the machine after this point either. I noted that my memory usage seemed to be fine via iStat (less than 50% utilization), but there may have been an unusually high memory pressure, but I could be wrong. Unless the kernel memory is inherently limited, I don't think an unusual amount of memory was allocated.

For context, here is the layout of the disks and pools:

```
/dev/disk0 (Macintosh HD):
  pool0.sparsebundle -> zpool pool0
/dev/disk1 (USB3 external):
  disk6.img -> /dev/disk6
  pool1.sparsebundle -> zpool pool1

pool0:
  large files
pool1:
  2tb zvol -> /dev/disk5
```

When I noticed that zfs had started to become unresponsive (and the whole system soon after), I was in the process of comparing disk5 (zvol on pool1) to disk6 (image on disk1) using dd and cmp (`dd if=/dev/disk5 | pv | cmp /dev/disk6`). This process had been running for over 18 hours without issue, but then I began to move ~90gb of regular files from pool0 to disk1 in two separate transfers via Finder which called DesktopServicesHelper for the actual move.

Below is the spindump filtered for what I thought was important. Note that I trimmed a couple thousand lines of identical threads from kernel_task that looked like this:

```
  Thread 0xb6a1d      1000 samples (1-1000) priority 81
 *1000  call_continuation + 23 (kernel + 2168551) [0xffffff80004116e7]
   *1000  txg_sync_thread + 571 (zfs + 342877) [0xffffff7f80b00b5d]
     *1000  spl_cv_timedwait + 260 (spl + 4501) [0xffffff7f80a77195]
       *1000  msleep + 98 (kernel + 6139570) [0xffffff80007daeb2]
         *1000  ??? (kernel + 6140292) [0xffffff80007db184]
           *1000  lck_mtx_sleep_deadline + 149 (kernel + 1272661) [0xffffff8000336b55]
             *1000  thread_block_reason + 175 (kernel + 1315119) [0xffffff800034112f]
               *1000  ??? (kernel + 1325012) [0xffffff80003437d4]
                 *1000  machine_switch_context + 367 (kernel + 2172719) [0xffffff800041272f]
```

The remainder of the report shows the dumps for the processes cmp, dd, DesktopServicesHelper, diskimages-helper, UserEventAgent, zed, several instances of zfs, zpool, and kernel_task. These were all of the processes interacting with the zfs kext at the time. I think all of the zfs processes were spawned as a result of GeekTool, the program I'm using to monitor the status of the zfs pools. It executes an info script every 15 seconds and whenever it times out, it usually means the system is about to lock up (if it hasn't already).

I don't think I should include the filtered spindump inline, so here's the gist of it (ha, get it?):

https://gist.github.com/vmrob/9e60a59bc169872d4e47

**Comment by ilovezfs** [comment]
Let us know if you still see the issue with https://www.dropbox.com/s/1c92xhxz28cacp3/yos-cef7965.pkg

**Comment by lundman** [comment]
Points of interest in the spindump are

```
                                         *1000  vnode_create + 814 (kernel + 3405662) [0xffffff800053f75e]
                                           *1000  ??? (kernel + 3432644) [0xffffff80005460c4]
                                             *1000  ??? (kernel + 3433434) [0xffffff80005463da]
                                               *1000  ??? (kernel + 3434736) [0xffffff80005468f0]
                                                 *1000  zfs_vnop_reclaim + 278 (zfs + 568237) [0xffffff7f80b37bad]
```
Because we are inside `vnode_create`, it means VFS is locked, so any other calls to VFS will wait. That means new vnodes, reclaims etc.

But this thread is correct, in that it should be able to call `zfs_rmnode` so the question is now about who is making zio_wait stall.

**Comment by vmrob** [comment]
@ilovezfs, the provided package failed installation. It seems that the kext couldn't be loaded.

From install.log:

```
Nov  4 10:16:26 rMBP-dev.local installd[12268]: ./postinstall: Loading spl.kext...
Nov  4 10:16:27 rMBP-dev.local installd[12268]: ./postinstall: /Library/Extensions/spl.kext failed to load - (libkern/kext) link error; check the system/kernel logs for errors or try kextutil(8).
Nov  4 10:16:27 rMBP-dev.local installd[12268]: ./postinstall: Loading zfs.kext...
Nov  4 10:16:27 rMBP-dev.local installd[12268]: ./postinstall: /Library/Extensions/zfs.kext failed to load - (libkern/kext) dependency load failed; check the system/kernel logs for errors or try kextutil(8).
Nov  4 10:16:27 rMBP-dev.local installd[12268]: ./postinstall: Died at /tmp/PKInstallSandbox.fouNrA/Scripts/net.lundman.openzfs.1010.pkg.0vYqZq/postinstall line 43.
```

From system.log:

```
Nov  4 09:53:46 rMBP-dev kernel[0]: ZFS: Attempting to unload ...
Nov  4 09:53:47 rMBP-dev kernel[0]: ZFS: Unloaded module v1.4.5-1
Nov  4 09:53:47 rMBP-dev kernel[0]: ZFS: Unloaded module
Nov  4 09:53:47 rMBP-dev kernel[0]: SPL: stop memory monitor
Nov  4 09:53:47 rMBP-dev kernel[0]: SPL: bsd_untimeout
Nov  4 09:53:47 rMBP-dev kernel[0]: SPL: wait for taskqs to empty
Nov  4 09:53:47 rMBP-dev kernel[0]: SPL: destroy taskq
Nov  4 09:53:48 rMBP-dev kernel[0]: SPL: Released 16 slabs
Nov  4 09:53:48 rMBP-dev kernel[0]: SPL: Released 524288 bytes from vmem_seg_arena
Nov  4 09:53:48 rMBP-dev kernel[0]: SPL: Unloaded module v1.4.5-1 (os_mem_alloc: 0) 
Nov  4 09:53:48 rMBP-dev com.apple.xpc.launchd[1] (org.openzfsonosx.zed[58]): Service exited with abnormal code: 1
Nov  4 09:53:48 rMBP-dev com.apple.xpc.launchd[1] (org.openzfsonosx.zed[82494]): Could not find and/or execute program specified by service: 2: No such file or directory: /usr/sbin/zed
Nov  4 09:53:48 rMBP-dev com.apple.xpc.launchd[1] (org.openzfsonosx.zed[82494]): Service setup event to handle failure and will not launch until it fires.
Nov  4 09:53:48 rMBP-dev.local com.apple.kextd[46]: Cache file /System/Library/Caches/com.apple.kext.caches/Directories/System/Library/Extensions/KextIdentifiers.plist.gz is out of date; not using.
Nov  4 09:53:49 rMBP-dev.local com.apple.kextd[46]: Cache file /System/Library/Caches/com.apple.kext.caches/Directories/Library/Extensions/KextIdentifiers.plist.gz is out of date; not using.
Nov  4 09:53:50 rMBP-dev kernel[0]: Refusing new kext net.lundman.kernel.dependencies.1.4.5, v12.5: a loaded copy with a different executable UUID is already present.
Nov  4 09:53:50 rMBP-dev kernel[0]: kxld[net.lundman.spl]: The following symbols are unresolved for this kext:
Nov  4 09:53:50 rMBP-dev kernel[0]: kxld[net.lundman.spl]: 	_VNOP_LOOKUP
Nov  4 09:53:50 rMBP-dev kernel[0]: Can't load kext net.lundman.spl - link failed.
Nov  4 09:53:50 rMBP-dev kernel[0]: Failed to load executable for kext net.lundman.spl.
Nov  4 09:53:50 rMBP-dev kernel[0]: Kext net.lundman.spl failed to load (0xdc008016).
Nov  4 09:53:50 rMBP-dev kernel[0]: Dependency net.lundman.spl of kext net.lundman.zfs failed to load.
Nov  4 09:53:50 rMBP-dev kernel[0]: Kext net.lundman.zfs failed to load (0xdc008015).
Nov  4 09:53:50 rMBP-dev kernel[0]: Failed to load kext net.lundman.zfs (error 0xdc008015).
Nov  4 09:53:50 rMBP-dev kernel[0]: Kext load request buffer from user space still retained by a kext; probable memory leak.
Nov  4 09:53:50 rMBP-dev.local com.apple.kextd[46]: Failed to load zfs.kext - (libkern/kext) dependency load failed.
```

**Comment by vmrob** [comment]
Restarting my machine solved that issue. I'll report back with any degradation of performance as it comes up.

**Comment by vmrob** [comment]
It seems that the provided package doesn't fix the issue. I'm sitting on a half-dead system right now, but don't have an elevated shell I can run spindump with. I'll be sure to keep one of those open next time.

**Comment by ilovezfs** [comment]
Do you know if this happens if you give ZFS real, whole external disks not disk images and sparsebundles on top of HFS+?

**Comment by vmrob** [comment]
@ilovezfs I do not know this. It would obviously be preferable to test that scenario and if I'm ever able to do that, I'll do so. I believe this issue to certainly be related to usage as my system has been up for over a week now, but only very light usage of the zvols.

For the time being, I'll attempt to come up with a reproducible test case using something similar to my current setup as I'm sure that will help much more that this bug report does currently.

**Comment by vmrob** [comment]
I wasn't able to come up with anything that consistently deadlocked despite running some pretty heavy reads and writes over smaller volumes, so I'm definitely not sure what the cause is, but I did have a deadlock today and was able to get a spindump. Once again, these are the processes with the zfs kexts loaded. I also trimmed several hundred similar threads from kernel_task as before.

https://gist.github.com/vmrob/2e9a90592734006d9e0a

I had a VM with the primary disk on a zvol doing some work--nothing heavy--and then noticed that Mission Control stopped functioning. I attempted to open Activity Monitor to investigate, but wasn't successful so I killed it. From there, I took a spindump and the proceeded to have everything else freeze up. I think I got it pretty early so hopefully this report is a little more valuable.
